### PR TITLE
NMS-13273: Add docs about var-bind in notifications

### PR DESCRIPTION
**Pull request by mfuhrmann**
### All Contributors

* [x] Have you read our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [x] Have you (electronically) signed the [OpenNMS Contributor Agreement](https://cla-assistant.io/OpenNMS/opennms)?

### Contribution Checklist

* Please [make an issue in the OpenNMS issue tracker](https://opennms.atlassian.net/) if there isn't one already.<br />Once there is an issue, please:
  1. update the title of this PR to be in the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the Jira link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
  4. once you've created this PR, please link to it in a comment in the Jira issue
  Don't worry if this sounds like a lot, we can help you get things set up properly.
* **If this code is likely to affect the UI, did you name your branch with `-smoke` in it to trigger smoke tests?**
* If this is a new or updated feature, is there documentation for the new behavior?
* If this is new code, are there unit and/or integration tests?
* If this PR targets a `foundation-*` branch, does it try to avoid changing files in `$OPENNMS_HOME/etc/`?

### What's Next?

A PR should be assigned at least 2 reviewers.  If you know that someone would be a good person to review your code, feel free to add them.

If you need help making additions or changes to the documentation related to your changes, please let us know.

In any case, if anything is unclear or you want help getting your PR ready for merge, please don't hesitate to say something in the comments here,
or in [the #opennms-development chat channel](https://chat.opennms.com/opennms/channels/opennms-development).

Once reviewer(s) accept the PR and the branch passes continuous integration, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the OGP) you are welcome to merge the PR when you're ready.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* Jira (Issue Tracker): https://opennms.atlassian.net/browse/NMS-13273

### CODE REVIEW

File: docs/modules/operation/pages/deep-dive/notifications/concepts.adoc [hunk 1]
```diff
@@ -94,3 +94,20 @@ This switch specifies the logic used when deciding whether to send a notice for
 If the switch is `off`, notices will never be sent to the target.
 If the switch is `on`, notices will always be sent to the target.
 If the switch is set to `auto`, the system deploys heuristics aimed at "doing the right thing."
+
+== Variable bindings
```
Comment (indigo423) [inline docs/modules/operation/pages/deep-dive/notifications/concepts.adoc:98]: This is a perfect use case for this functionality: https://chat.opennms.com/opennms/pl/gnyui5sgwiddpfi1moymrrj8ry

**Review by indigo423 — COMMENTED** [review]
I would rewrite it against this use case, I think having people looking for something like this is more likely.

File: docs/modules/operation/pages/deep-dive/notifications/concepts.adoc (hunk 1 above)
Comment (mmahacek) [reply docs/modules/operation/pages/deep-dive/notifications/concepts.adoc:98]: Thanks for this feedback.  Honestly, I've never used either of those settings on notifications, so this doc update is welcomed.

Comment (mmahacek) [reply docs/modules/operation/pages/deep-dive/notifications/concepts.adoc]: According to the XSD, it is 1 varbind and many parameters
```
        <element maxOccurs="unbounded" minOccurs="0" ref="this:parameter"/>
        <element maxOccurs="1" minOccurs="0" ref="this:varbind"/>
```

**Review by mmahacek — COMMENTED** [review]
I'm going to propose adding a tweak to the web UI to indicate that the varbind and parameter options act as event filters.  This has never been clear to me what those sections are for, and having a note in the web UI would benefit users.

**Comment by mfuhrmann** [comment]
@mmahacek @indigo423 I've added the BSM example. Two questions here:

1. Do we want to start to have example code somewhere else? Similar to the references? I think we always wanted to separate this, no? 
2. Isn't the `vbname` always an event parameter? If so, I am wondering if we should call like this in these descriptions: https://github.com/OpenNMS/opennms/pull/6955/files#diff-c8d6cc2c1f5ad9d047ace92b42d00f237d91c67e60bf1aa7b49cfed21f144fc2R124 Or can a varbind also something else than an event parameter?

File: docs/modules/operation/pages/deep-dive/notifications/concepts.adoc [hunk 2]
```diff
@@ -94,3 +94,55 @@ This switch specifies the logic used when deciding whether to send a notice for
...
+
+== Variable bindings filters
+
+Event parameters can additionally act as a filter method.
+This can be helpful for events that don't have an association to a specific node and the common filters only match against interfaces.
+By defining the `vbname` and `vbvalue`, the notification will only be sent with if the event parameter matches the given value.
+
+Example 1:
+
+In this example, notifications on a SNMP Link Down event would be suppressed for trunk ports.
```
Comment (mmahacek) [inline docs/modules/operation/pages/deep-dive/notifications/concepts.adoc:106]: Wouldn't this be the opposite, only triggering events that do match `*Trunk*`?

File: docs/modules/operation/pages/deep-dive/notifications/concepts.adoc [hunk 3]
```diff
@@ -94,3 +94,55 @@ This switch specifies the logic used when deciding whether to send a notice for
...
+         <vbvalue>~^.*Trunk.*</vbvalue> <2>
+      </varbind>
+   </notification>
+----
+<1> Varbind or parameter name to filter
+<2> Value to match for the specified varbind/parameter.
+If the value starts with a `~`, it will be processed as a regular expression, otherwise it will be a string match.
+
+Example 2:
+
```
Comment (mmahacek) [inline docs/modules/operation/pages/deep-dive/notifications/concepts.adoc:129]:
```suggestion

In this example, Business Service Monitoring problem events would only trigger notifications if there is a parameter match with the specified business service name pattern.

```

**Comment by mmahacek** [comment]
> @mmahacek @indigo423 I've added the BSM example. Two questions here:
> 
> 1. Do we want to start to have example code somewhere else? Similar to the references? I think we always wanted to separate this, no?
> 2. Isn't the `vbname` always an event parameter? If so, I am wondering if we should call like this in these descriptions: https://github.com/OpenNMS/opennms/pull/6955/files#diff-c8d6cc2c1f5ad9d047ace92b42d00f237d91c67e60bf1aa7b49cfed21f144fc2R124 Or can a varbind also something else than an event parameter?

1. It's best to keep the example snippets near the section that discusses the feature, so it's easier to keep them in context.
2. I don't know the answer as I haven't used this feature enough.

**Comment by mfuhrmann** [comment]
> * Do we want to start to have example code somewhere else? Similar to the references? I think we always wanted to separate this, no?

Ah. See here: https://opennms.atlassian.net/browse/NMS-13555
I would like to start on this which affects also this PR I think. I would suggest to create a sub page in `references` called `examples`. Similar to the monitors/detectors etc. Wdyt?

File: docs/modules/operation/pages/deep-dive/notifications/concepts.adoc (hunk 2 above)
Comment (mfuhrmann) [reply docs/modules/operation/pages/deep-dive/notifications/concepts.adoc:106]: Yes. It is the positive match.

```suggestion
In this example, notifications would only be triggered on a SNMP Link Down events that contain the string `Trunk`.
```

**Comment by mfuhrmann** [comment]
> * It's best to keep the example snippets near the section that discusses the feature, so it's easier to keep them in context.

@mmahacek Why do we want to do this in the poller/detector sections, but not here? :thinking: Just asking. Do you think there is a difference? I mean, we would link to the corresponding examples.

**Comment by mmahacek** [comment]
> @mmahacek Why do we want to do this in the poller/detector sections, but not here? 🤔 Just asking. Do you think there is a difference? I mean, we would link to the corresponding examples.

I don't have a personal desire to make any changes to poller/collector/detector documentation, except to add any missing data or correct any found inaccuracies.

**Comment by mfuhrmann** [comment]
> * Do we want to start to have example code somewhere else? Similar to the references? I think we always wanted to separate this, no?
> * Isn't the `vbname` always an event parameter? If so, I am wondering if we should call like this in these descriptions: https://github.com/OpenNMS/opennms/pull/6955/files#diff-c8d6cc2c1f5ad9d047ace92b42d00f237d91c67e60bf1aa7b49cfed21f144fc2R124 Or can a varbind also something else than an event parameter?

1. We decided to not move anything out and let the examples here.
2. I had a talk with @cgorantla and to understand those `varbind`s better. It is indeed just another name for `event parameter`. So let's rename it in the docs.

File: docs/modules/operation/pages/deep-dive/notifications/concepts.adoc [hunk 4]
```diff
@@ -94,3 +94,57 @@ This switch specifies the logic used when deciding whether to send a notice for
...
+      <text-message>%descr%</text-message>
+      <subject>Notice #%noticeid% - Link Down</subject>
+      <numeric-message>111-%noticeid%</numeric-message>
+      <varbind>
+         <vbname>ifAlias</vbname> <1>
+         <vbvalue>~^.*Trunk.*</vbvalue> <2>
+      </varbind>
+   </notification>
+----
+<1> Varbind or parameter name to filter
```
Comment (mfuhrmann) [inline docs/modules/operation/pages/deep-dive/notifications/concepts.adoc:124]:
```suggestion
<1> Event parameter name to filter
```

File: docs/modules/operation/pages/deep-dive/notifications/concepts.adoc [hunk 5]
```diff
@@ -94,3 +94,57 @@ This switch specifies the logic used when deciding whether to send a notice for
...
+<2> Value to match for the specified varbind/parameter.
```
Comment (mfuhrmann) [inline docs/modules/operation/pages/deep-dive/notifications/concepts.adoc:125]:
```suggestion
<2> Value to match for the specified event parameter.
```

File: docs/modules/operation/pages/deep-dive/notifications/concepts.adoc [hunk 6]
```diff
@@ -94,3 +94,57 @@ This switch specifies the logic used when deciding whether to send a notice for
...
+      <text-message>BSM Root Cause Analysis: %parm[rootCause]%</text-message>
+      <subject>%parm[businessServiceName]% is impacted</subject>
+      <numeric-message>111-%noticeid%</numeric-message>
+      <varbind>
+         <vbname>businessServiceName</vbname> <1>
+         <vbvalue>~^MailCluster.*</vbvalue> <2>
+      </varbind>
+   </notification>
+----
+<1> Varbind or parameter name to filter
```
Comment (mfuhrmann) [inline docs/modules/operation/pages/deep-dive/notifications/concepts.adoc:148]:
```suggestion
<1> Event parameter name to filter
```

File: docs/modules/operation/pages/deep-dive/notifications/concepts.adoc [hunk 7]
```diff
@@ -94,3 +94,57 @@ This switch specifies the logic used when deciding whether to send a notice for
...
+<2> Value to match for the specified varbind/parameter.
```
Comment (mfuhrmann) [inline docs/modules/operation/pages/deep-dive/notifications/concepts.adoc:149]:
```suggestion
<2> Value to match for the specified event parameter.
```

File: docs/modules/operation/pages/deep-dive/notifications/concepts.adoc [hunk 8]
```diff
@@ -94,3 +94,57 @@ This switch specifies the logic used when deciding whether to send a notice for
 If the switch is `off`, notices will never be sent to the target.
 If the switch is `on`, notices will always be sent to the target.
 If the switch is set to `auto`, the system deploys heuristics aimed at "doing the right thing."
+
+== Variable bindings filters
+
+Event parameters can additionally act as a filter method.
```
Comment (mmahacek) [inline docs/modules/operation/pages/deep-dive/notifications/concepts.adoc:100]:
```suggestion
Event parameters can be used to filter which events trigger a notification.
```

File: docs/modules/operation/pages/deep-dive/notifications/concepts.adoc [hunk 9]
```diff
@@ -94,3 +94,57 @@ This switch specifies the logic used when deciding whether to send a notice for
...
+This can be helpful for events that don't have an association to a specific node and the common filters only match against interfaces.
```
Comment (mmahacek) [inline docs/modules/operation/pages/deep-dive/notifications/concepts.adoc:101]:
```suggestion
```

**Review by mmahacek — COMMENTED** [review]
Couple extra wording tweaks

**Review by mmahacek — APPROVED** [review]

**Pull request merged**